Incorrect `DataCorruptionException` thrown upon recovery

**Problem description**
This should not happen:
```
2018-06-14 12:58:18,120 6767489 [core-15] ERROR i.p.s.server.logs.DurableLog - DurableLog[0] Recovery FAILED.
io.pravega.segmentstore.server.DataCorruptionException: Deserialization failed.
	at io.pravega.segmentstore.server.logs.DataFrameReader.getNext(DataFrameReader.java:108)
	at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverAllOperations(RecoveryProcessor.java:159)
	at io.pravega.segmentstore.server.logs.RecoveryProcessor.performRecovery(RecoveryProcessor.java:93)
	at io.pravega.segmentstore.server.logs.DurableLog.performRecovery(DurableLog.java:217)
Caused by: java.io.IOException: io.pravega.segmentstore.storage.DurableDataLogException: Error while reading from BookKeeper.
	at io.pravega.segmentstore.server.logs.DataFrameInputStream.fetchNextEntry(DataFrameInputStream.java:197)
	at io.pravega.segmentstore.server.logs.DataFrameInputStream.beginRecord(DataFrameInputStream.java:140)
	at io.pravega.segmentstore.server.logs.DataFrameReader.getNext(DataFrameReader.java:86)
	... 11 common frames omitted
Caused by: io.pravega.segmentstore.storage.DurableDataLogException: Error while reading from BookKeeper.
	at io.pravega.segmentstore.storage.impl.bookkeeper.LogReader.openNextLedger(LogReader.java:154)
	at io.pravega.segmentstore.storage.impl.bookkeeper.LogReader.getNext(LogReader.java:100)
	at io.pravega.segmentstore.storage.impl.bookkeeper.LogReader.getNext(LogReader.java:34)
	at io.pravega.segmentstore.server.logs.DataFrameInputStream.getNextFrame(DataFrameInputStream.java:292)
	at io.pravega.segmentstore.server.logs.DataFrameInputStream.getNextFrameEntry(DataFrameInputStream.java:272)
	at io.pravega.segmentstore.server.logs.DataFrameInputStream.fetchNextEntry(DataFrameInputStream.java:195)
	... 13 common frames omitted
Caused by: org.apache.bookkeeper.client.BKException$BKTimeoutException: null
	at org.apache.bookkeeper.client.SynchCallbackUtils.finish(SynchCallbackUtils.java:63)
	at org.apache.bookkeeper.client.LedgerHandle$SyncReadCallback.readComplete(LedgerHandle.java:1801)
	at org.apache.bookkeeper.client.PendingReadOp.submitCallback(PendingReadOp.java:568)
	at org.apache.bookkeeper.client.PendingReadOp$LedgerEntryRequest.fail(PendingReadOp.java:156)
	at org.apache.bookkeeper.client.PendingReadOp$SequenceReadRequest.sendNextRead(PendingReadOp.java:383)
	at org.apache.bookkeeper.client.PendingReadOp$SequenceReadRequest.logErrorAndReattemptRead(PendingReadOp.java:416)
	at org.apache.bookkeeper.client.PendingReadOp.readEntryComplete(PendingReadOp.java:522)
	at org.apache.bookkeeper.proto.PerChannelBookieClient$ReadCompletion$1.readEntryComplete(PerChannelBookieClient.java:1780)
	at org.apache.bookkeeper.proto.PerChannelBookieClient$9.safeRun(PerChannelBookieClient.java:1013)
	at org.apache.bookkeeper.util.SafeRunnable.run(SafeRunnable.java:33)
	... 7 common frames omitted
```

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameInputStream.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameInputStream.java
@@ -20,6 +20,7 @@ import java.io.IOException;
 import java.io.InputStream;
 import javax.annotation.concurrent.NotThreadSafe;
 import lombok.Getter;
+import lombok.SneakyThrows;
 import lombok.extern.slf4j.Slf4j;
 
 /**
@@ -74,6 +75,7 @@ public class DataFrameInputStream extends InputStream {
     }
 
     @Override
+    @SneakyThrows(DurableDataLogException.class)
     public int read() throws IOException {
         while (!this.closed) {
             int r = this.currentEntry.getData().read();
@@ -92,6 +94,7 @@ public class DataFrameInputStream extends InputStream {
     }
 
     @Override
+    @SneakyThrows(DurableDataLogException.class)
     public int read(byte[] buffer, int index, int length) throws IOException {
         Preconditions.checkNotNull(buffer, "buffer");
         if (index < 0 || length < 0 | index + length > buffer.length) {
@@ -130,8 +133,9 @@ public class DataFrameInputStream extends InputStream {
      * @return True if a new record can be read, false if we reached the end of the DataFrameInputStream and can no longer
      * read.
      * @throws IOException If an IO Exception occurred.
+     * @throws DurableDataLogException If a non-IO Exception has occurred, usually thrown by the underlying DurableDataLog.
      */
-    boolean beginRecord() throws IOException {
+    boolean beginRecord() throws IOException, DurableDataLogException {
         try {
             if (this.currentEntry != null && !this.prefetchedEntry) {
                 endRecord();
@@ -153,8 +157,9 @@ public class DataFrameInputStream extends InputStream {
      *
      * @return A RecordInfo containing metadata about the record that just ended, such as addressing information.
      * @throws IOException If an IO Exception occurred.
+     * @throws DurableDataLogException If a non-IO Exception has occurred, usually thrown by the underlying DurableDataLog.
      */
-    DataFrameRecord.RecordInfo endRecord() throws IOException {
+    DataFrameRecord.RecordInfo endRecord() throws IOException, DurableDataLogException {
         DataFrameRecord.RecordInfo r = this.currentRecordBuilder.build();
         while (this.currentEntry != null) {
             if (this.currentEntry.isLastRecordEntry()) {
@@ -181,7 +186,7 @@ public class DataFrameInputStream extends InputStream {
         this.prefetchedEntry = false;
     }
 
-    private void fetchNextEntry() throws IOException {
+    private void fetchNextEntry() throws IOException, DurableDataLogException {
         Exceptions.checkNotClosed(this.closed, this);
         if (this.prefetchedEntry) {
             assert this.currentEntry != null : "prefetchEntry==true, but currentEntry==null";
@@ -190,12 +195,7 @@ public class DataFrameInputStream extends InputStream {
         }
 
         while (!this.closed) {
-            DataFrame.DataFrameEntry nextEntry;
-            try {
-                nextEntry = getNextFrameEntry();
-            } catch (DurableDataLogException ex) {
-                throw new IOException(ex);
-            }
+            DataFrame.DataFrameEntry nextEntry = getNextFrameEntry();
 
             if (nextEntry == null) {
                 // 'null' means no more entries (or frames). Since we are still in the while loop, it means we were in the

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameReader.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameReader.java
@@ -77,7 +77,7 @@ class DataFrameReader<T extends SequencedItemList.Element> implements CloseableI
      * @return A DataFrameRecord with the requested operation. If no more Operations are available, null is returned.
      */
     @Override
-    public DataFrameRecord<T> getNext() throws DataCorruptionException {
+    public DataFrameRecord<T> getNext() throws DataCorruptionException, DurableDataLogException {
         Exceptions.checkNotClosed(this.closed, closed);
 
         try {

--- a/test/testcommon/src/main/java/io/pravega/test/common/ErrorInjector.java
+++ b/test/testcommon/src/main/java/io/pravega/test/common/ErrorInjector.java
@@ -10,10 +10,10 @@
 package io.pravega.test.common;
 
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.CompletionException;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Predicate;
 import java.util.function.Supplier;
+import lombok.SneakyThrows;
 
 /**
  * Simple count-based error injector.
@@ -56,11 +56,14 @@ public class ErrorInjector<T extends Throwable> {
      * @param injector The Error Injector to use.
      * @param <T>      The type of exception to throw.
      */
+    @SneakyThrows
     public static <T extends Throwable> void throwSyncExceptionIfNeeded(ErrorInjector<T> injector) {
         if (injector != null) {
             T ex = injector.generateExceptionIfNecessary();
             if (ex != null) {
-                throw new CompletionException(ex);
+                // Throw the exception directly (thanks to SneakyThrows), vs wrapping it in some other exception, which
+                // may not be handled or expected by the "catching" code.
+                throw ex;
             }
         }
     }

